torch/csrc/jit/python/python_sugared_value.cpp

```diff
@@ -226,11 +226,19 @@ std::shared_ptr<SugaredValue> CUDAPythonModuleValue::attr(
   const std::unordered_set<std::string> cuda_ops = {
       "current_stream",
       "default_stream",
-      "_current_device",
-      "_set_device",
+      "_cuda_setDevice",
+      "_cuda_setStream",
       "device_index",
       "device_count",
-      "set_stream"};
+  };
+
+  // current_device is treated as an aten operator here. This is done
+  // in order to differentiate from c10::cuda::current_device.
+  // We cannot re-name current_device , since this needs to be consistent with
+  // eager mode naming torch.cuda.current_device.
+  if (field == "current_device") {
+    return std::make_shared<BuiltinFunction>(Symbol::aten(field), c10::nullopt);
```

Comment:
The rationale behind making `current_device` as an aten operator:
- `torch.cuda.current_device` is used quite frequently in the eager tests/ JIT tests. So I assumed, most users would use this API quite frequently and hence changing this API for JIT would not be that comfortable for the user. These are my thoughts, but I am not sure how much importance these hold. 

- After my discussion with Pyper(Xiaodang), he wanted the JIT APIs to be as close to eager APIs as possible, this way the changes for them to move to JIT would be minimal and also prevent code duplication. 

>>better than putting this op in the aten namespace because the implementation only works for CUDA.
      - I have moved the `aten::current_device` under `ifdef`. So this will take care that `current_device` is only available for `CUDA` builds. 

One more option which I can think of is, we can try using a different namespace like `_cuda`. Wdyt?

Reply:
> `torch.cuda.current_device` is used quite frequently in the eager tests/ JIT tests. So I assumed, most users would use this API quite frequently and hence changing this API for JIT would not be that comfortable for the user. 

Do you mean `torch.current_device`? `torch.x` is currently desugared to `aten::x`, and that's the only thing I can think of that might require `current_device` to reside in the JIT `aten` namespace. `torch.cuda.current_device` should be resolvable using the `CUDAPythonModuleValue` class you added.

> - I have moved the aten::current_device under ifdef. So this will take care that current_device is only
available for CUDA builds.

The fact that it is in the `aten` namespace gives the impression that it is not specific to CUDA devices. Whether it is available in builds with or without CUDA is not relevant.

Reply:
There is a bit to untangle here.

>  I have moved the aten::current_device under ifdef. So this will take care that current_device is only
available for CUDA builds.

Although we have some APIs that only work under CUDA, our general philosophy is to always provide the APIs, and give a good error when people try to use them in a CUDA-less environment. So for example, I can run this on my OS X (with no GPU) and get a reasonable error:

```
>>> import torch
>>> torch.cuda.current_device()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ezyang/miniconda3/lib/python3.7/site-packages/torch/cuda/__init__.py", line 366, in current_device
    _lazy_init()
  File "/Users/ezyang/miniconda3/lib/python3.7/site-packages/torch/cuda/__init__.py", line 166, in _lazy_init
    raise AssertionError("Torch not compiled with CUDA enabled")
AssertionError: Torch not compiled with CUDA enabled
```

I'm not exactly sure what you did with ifdefs, but if this causes code to stop compiling if you run it on  a CPU-only build of PyTorch, the old code was probably better.

> torch.cuda.current_device is used quite frequently in the eager tests/ JIT tests. So I assumed, most users would use this API quite frequently and hence changing this API for JIT would not be that comfortable for the user. These are my thoughts, but I am not sure how much importance these hold.

This is all right. But you then seem to implicitly imply "and therefore the operator name may not change". But there is no *intrinsic* relationship between how an operator surfaces in Python (torch.cuda.current_device) and how it shows up in the IR (aten::current_device versus cuda::current_device, etc); as simple evidence, consider why the word "aten" doesn't show up anywhere in the C++ frontend. So yes, you shouldn't change the frontend API name... but that doesn't mean that you can't use the appropriate operator name. If it is not mechanically clear how to do this, shout.

Reply:
> One more option which I can think of is, we can try using a different namespace like _cuda.

If this helps deal with the problem where there is a conflict between the C++ function `c10::cuda::get_device` and `c10::cuda::get_device` the `c10::Symbol` instance, I think using a namespace named `_cuda` instead would work too.